service-base/src/main/java/org/eclipse/hono/service/limiting/ConnectionLimitManager.java

```diff
@@ -0,0 +1,92 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.hono.service.limiting;
+
+import org.eclipse.hono.config.ProtocolAdapterProperties;
+import org.eclipse.hono.service.metric.Metrics;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class manages a limit of concurrent connections. The limit can be configured in
+ * {@link ProtocolAdapterProperties#setMaxConnections(int)}. If no value is configured, it will set a limit based on the
+ * the given strategy.
+ */
+public class ConnectionLimitManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ConnectionLimitManager.class);
+
+    private final ConnectionLimitStrategy strategy;
+    private final Metrics metrics;
+    private final int limit;
+
+    /**
+     * Creates an instance configured with a limiting strategy, the adapter's metrics to get the current number of
+     * connections and the adapter's properties for the configured limit.
+     * <p>
+     * If the protocol adapter properties are null or no limit is configured, the recommended limit of the strategy will
+     * be set as the limit.
+     * 
+     * @param strategy The strategy to configure the connection limit.
+     * @param metrics The metrics of the adapter.
+     * @param config The configuration of the adapter or {@code null}.
+     */
+    public ConnectionLimitManager(final ConnectionLimitStrategy strategy, final Metrics metrics,
```

Comment:
I'd rather not tie this class to `Metrics`. In fact, we only ever need to read the current number of connections, right?
So, why not use something like
```
public ConnectionLimitManager(final ConnectionLimitStrategy strategy, final Supplier<Integer> currentConnections, final ProtocolAdapterProperties config)
```
instead.
A supplier based on Metrics could then be passed in like this:
```
new ConnectionLimitManager(strategy, () -> metrics.getNumberOfConnections(), config) ...
```
That way we could more easily swap out the source for the number of connections if necessary. WDYT?

Reply:
Yes, that is very good.